### ISSUE TITLE
Fitters do not respect pulse numbers

### ISSUE THREAD

**Issue by aarchiba** [nanograv/PINT#812]
While TOAs objects can keep track of pulse numbers, and Resduals objects can be set up to respect these with `track="use_pulse_numbers"`, there does not appear to be any way to get a Fitter to respect the pulse numbers on input TOAs. This makes it impossible to take advantage of an approximate initial timing solution to lock in pulse numbers and remove nasty discontinuities from the fitting process.
![bad-fitter](https://user-images.githubusercontent.com/136867/99291195-5471f980-2837-11eb-97b9-12ccf30cc0b8.png)

```python
import astropy.units as u
from astropy.visualization import quantity_support
import matplotlib.pyplot as plt
from pint.models import get_model
from pint.toa import get_TOAs
from pint.residuals import Residuals
import pint.fitter

quantity_support()

m = get_model('PINT/tests/datafile/B1855+09_NANOGrav_dfg+12_DMX.par')
t = get_TOAs('PINT/tests/datafile/B1855+09_NANOGrav_dfg+12.tim')
t.compute_pulse_numbers(m)
r = Residuals(t,m)
plt.plot(t.get_mjds(), r.time_resids.to(u.us), ".", label="good")
m.F0.value += 5e-9
r = Residuals(t,m)
plt.plot(t.get_mjds(), r.time_resids.to(u.us), ".", label="no pulse numbers")
r = Residuals(t,m,track_mode="use_pulse_numbers")
plt.plot(t.get_mjds(), r.time_resids.to(u.us), ".", label="pulse numbers")
for p in m.params:
    getattr(m,p).frozen=True
m.F0.frozen=False
f = pint.fitter.WLSFitter(t, m)
f.fit_toas()
plt.plot(t.get_mjds(), f.resids.time_resids.to(u.us), ".", label="fit")
plt.legend()
plt.savefig("bad-fitter.png")
```
produces the output
```
floatParameter( F0                  186.49408156698235146 (Hz) +/- 6.98911818e-12 Hz frozen=False)
floatParameter( F0                  186.49408157198235146 (Hz) +/- 6.98911818e-12 Hz frozen=False)
floatParameter( F0                  186.49408157510560433 (Hz) +/- 7.54283184026823e-14 Hz frozen=False)
```
In other words, the `F0` does not return to the original value.

**Comment by scottransom** [comment]
I'm slightly confused.  So the issue is with TOAs that have pulse numbers specified on input only?

**Comment by aarchiba** [comment]
You can plot residuals in a way that respects pulse numbers. But you can't *fit* them in a way that respects pulse numbers - the fitter should recover the original `F0` but the phase wrap makes it fail to do so.

**Comment by aarchiba** [comment]
Green dots show the residuals with pulse numbers locked in; if the Fitter used those residuals it would work fine. But Fitters make their own Residuals objects without respecting any `track=` setting (or defaulting to `use_pulse_numbers` if available) so the Fitter works with, effectively, the orange residuals, and can therefore do no better than the red  post-fit residuals.

**Comment by scottransom** [comment]
Ah, now I understand (I think).  That seems like it should be an easy fix by simply passing track= as a keyword param to the fitter, right?  (and then using that when the fitters create the residuals object).   By far the trickiest bit in all of this is getting the residuals objects to behave correctly and I *think* that happens.

**Comment by scottransom** [comment]
Note that you can pass a residuals instance to the Fitter().  If you create the residuals object with track=, does the fit then work?  I'm curious....

**Comment by scottransom** [comment]
Nevermind...  WLSFitter() incorrectly overwrites Fitter()'s __init__().  Anyway, I think this can be fairly easily fixed and we'll try to get to it soon.

**Comment by aarchiba** [comment]
> Note that you can pass a residuals instance to the Fitter(). If you create the residuals object with track=, does the fit then work? I'm curious....

Sadly not. That's just a shortcut for computing statistics on already-existing residuals, doesn't do anything for plain Fitters, and isn't supported by non GLSFitters.

**Comment by aarchiba** [comment]
> Nevermind... WLSFitter() incorrectly overwrites Fitter()'s **init**(). Anyway, I think this can be fairly easily fixed and we'll try to get to it soon.

PR running test suite now.

**Comment by aarchiba** [comment]
Also WLSFitter correctly uses super(); WidebandFitter does not, and I can't tell whether that is intentional or not.

**Comment by luojing1211** [comment]
@aarchiba That is a good catch. I believe there are a lot of divergences from the fitter base, the test did not fail. I will fix it.

**Comment by aarchiba** [comment]
The PR will appear once it passes the test suite on my machine. I've tidied the things I needed to to make that work.

**Comment by aarchiba** [comment]
I'm trying to speed this up but ran into bug #821 .

**Comment by aarchiba** [comment]
Also there is some confusion - `.par` files can contain `TRACK -2` or `TRACK 0` (other options not supported) and previously this overrode values explicitly passed to the `Residuals` constructor. This seems wrong so I made explicit values override `.par` file values. If the `.par` file does not contain a `TRACK` statement I'd prefer to default to using pulse numbers if present and not if missing, but there is also the question of what to do when some TOAs have pulse numbers and others do not (for example new data being added to an existing timing data set).

**Comment by poneill129** [comment]
Hi all, this bug was found while creating a tutorial notebook for PINT.  Many thanks to @aarchiba for compiling this bug report.
